oauth2_http/java/com/google/auth/oauth2/AwsCredentials.java

```diff
@@ -135,11 +146,38 @@ public AccessToken refreshAccessToken() throws IOException {
 
   @Override
   public String retrieveSubjectToken() throws IOException {
+    // AWS IDMSv2 introduced a requirement for a session token to be present
+    // with the requests made to metadata endpoints. This requirement is to help
+    // prevent SSRF attacks.
+    // Presence of "imdsv2_session_token_url" in Credential Source of config file
+    // will trigger a flow with session token, else there will not be a session
+    // token with the metadata requests.
+    // Both flows work for IDMS v1 and v2. But if IDMSv2 is enabled, then if
+    // session token is not present, Unauthorized exception will be thrown.
+    Map<String, Object> metadataHeaders = new HashMap<>();
+    if (awsCredentialSource.imdsv2SessionTokenUrl != null) {
+      Map<String, Object> tokenRequestHeaders =
+          new HashMap<String, Object>() {
+            {
+              put("x-aws-ec2-metadata-token-ttl-seconds", "300");
+            }
+          };
+
+      String imdsv2SessionToken =
+          retrieveResource(
+              awsCredentialSource.imdsv2SessionTokenUrl,
+              "Session Token",
+              HttpMethods.PUT,
+              tokenRequestHeaders,
+              /* content= */ null);
+      metadataHeaders.put("x-aws-ec2-metadata-token", imdsv2SessionToken);
+    }
```

Comment:
How about we  factor it out into a separate method like getRegion and getCredentials? It is an optional codepath and implementation details does not matter in this method.

Reply:
Yeah. I had done that way for other languages. Missed here.